### DISCUSSION TITLE
Question about padding for TLS 1.2 with provided AES-256-CBC cipher (OSSL_FUNC_cipher_update versus OSSL_FUNC_cipher_final)

### DISCUSSION THREAD

**Discussion by UponAPaintedOcean** [openssl/openssl#22548] [Q&A] [answered]
Hello,

I'm trying to implement a provider that can do TLS 1.2 with OpenSSL 3.0.7 and AES-256-CBC as cipher.

My question is about the invocation of EVP_CipherUpdate in tls1_enc, specifically about padding. At one point my provider gets 18 bytes as input. My implementation encrypts the first 16 bytes, but I don't see a corresponding EVP_CipherFinal invocation, so the remaining 2 bytes are effectively discarded and TLS fails overall.

tls1_enc has this comment:

```
* We only do this for legacy ciphers. Provided ciphers add the
* padding on the provider side.
```

My understanding is that normally only OSSL_FUNC_cipher_final should handle padding. But maybe in this case my implementation of OSSL_FUNC_cipher_update should perform padding too? If so, how can I distinguish this special case?

**Comment by mattcaswell** [comment]
> My understanding is that normally only OSSL_FUNC_cipher_final should handle padding.

For non-TLS based usage - yes.

> But maybe in this case my implementation of OSSL_FUNC_cipher_update should perform padding too? If so, how can I distinguish this special case?

In order to be usable in TLS your cipher needs to support the `OSSL_CIPHER_PARAM_TLS_VERSION` `EVP_CIPHER_CTX` parameter. See the "Settable EVP_CIPHER_CTX parameters" section of this man page:

https://www.openssl.org/docs/man3.0/man3/EVP_EncryptInit.html

Probably we could do with some doc updates there. The documentation is a little...err...terse! :-)

If the `OSSL_CIPHER_PARAM_TLS_VERSION` parameter has been passed then you know that the cipher is being used in a TLS context. With TLS each call to EVP_CipherUpdate() is passing a whole TLS record and TLS record padding is expected to be applied by the cipher. This is how the built-in OpenSSL ciphers do it:

https://github.com/openssl/openssl/blob/bdac6bcc50bb93b61989cee3f74a65c34abe1bd2/providers/implementations/ciphers/ciphercommon.c#L252-L326

**Comment by mattcaswell** [comment]
Oh....I should also mention that it needs to support the gettable `OSSL_CIPHER_PARAM_TLS_MAC` param and settable `OSSL_CIPHER_PARAM_TLS_MAC_SIZE` param.

If the mac size > 0 then the cipher needs to also be provide a pointer to the mac in a decrypted record in order to support mac-then-encrypt (i.e. if the encrypt-then-mac extension is *not* negotiated).

See:

https://github.com/openssl/openssl/blob/bdac6bcc50bb93b61989cee3f74a65c34abe1bd2/providers/implementations/ciphers/ciphercommon_block.c#L137-L190

and

https://github.com/openssl/openssl/blob/bdac6bcc50bb93b61989cee3f74a65c34abe1bd2/ssl/record/tls_pad.c#L96-L178

This is quite tricky to do correctly since it needs to be done in constant time to avoid certain attacks.

**Comment by UponAPaintedOcean** [comment]
Thank you very much. Your explanation has been very helpful. I'm getting much further now.

**Comment by UponAPaintedOcean** [comment]
I still have a question, specifically about decryption, which works differently than I expected.

For example, suppose my OSSL_FUNC_cipher_update, after the removal of padding, has 51 bytes of output, meaning 16 bytes random IV and 35 bytes of payload. To get the whole handshake to work, I find that I have to copy back all 51 bytes to *out, but I have to set *outl to 35, not 51.

I could very well be misunderstanding, but this seems to be because of this fragment in tls1_enc in https://github.com/openssl/openssl/blob/openssl-3.0.7/ssl/record/ssl3_record.c :

https://github.com/openssl/openssl/blob/19cc035b6c6f2283573d29c7ea7f7d675cf750ce/ssl/record/ssl3_record.c#L1196-L1200

The intention of this code is to skip the 16 bytes random (bs == 16). I would expect it to subtract bs also from recs[0].length. But because it does not, my workaround is basically to compensate for that by assigning a 16 bytes smaller value to *outl. Everything works now, but it feels wrong to do this.

For what it's worth: on the corresponding legacy cipher path a little further down in tls1_enc, it does subtract bs also from recs[ctr].length:

https://github.com/openssl/openssl/blob/19cc035b6c6f2283573d29c7ea7f7d675cf750ce/ssl/record/ssl3_record.c#L1246-L1253

All this is still with OpenSSL 3.0.7. I haven't had time yet to try newer versions. The code in master (https://github.com/openssl/openssl/blob/master/ssl/record/methods/tls1_meth.c) looks the same in this regard, though.

**Comment by UponAPaintedOcean** [comment, marked as answer]
I looked into this a little more. Using the default provider instead of my own, I end up here:

https://github.com/openssl/openssl/blob/19cc035b6c6f2283573d29c7ea7f7d675cf750ce/providers/implementations/ciphers/ciphercommon_block.c#L176-L184

The addition of blocksize to buf only has a local effect, whereas the addition to *buflen ultimately affects the value that function tls1_enc sees.

It feels somewhat incongruous to me, but at least this explains it and I can work with it.
